src/libstrongswan/plugins/botan/botan_util_keys.c

```diff
@@ -128,6 +130,18 @@ public_key_t *botan_public_key_load(key_type_t type, va_list args)
 		this = botan_ed_public_key_adopt(pubkey);
 	}
 	else
+#endif
+#ifdef BOTAN_HAS_ML_DSA
+	if (streq(name, "ML-DSA") && (type == KEY_ANY || type == KEY_ML_DSA_44 ||
+							  type == KEY_ML_DSA_65 || type == KEY_ML_DSA_87))
+	{
+		if (type == KEY_ANY)
+		{
+			type = public_key_info_decode(blob, NULL);
```

Comment:
What happens if the type can't be determined? `botan_ml_dsa_public_key_adopt()` should probably fail if `type` is `KEY_ANY`.

Reply:
botan parses the public key already in line 92 and exits if the key type is not supported:
```
	if (botan_pubkey_load(&pubkey, blob.ptr, blob.len))
	{
		botan_rng_destroy(rng);
		return NULL;
	}
```
Our problem is that the botan **`pubkey`** object currently doesn't offer a method to extract which type of ML-DSA key was loaded (according to the source code this is on the TODO list for botan 4). Therefore we have to do our own PKCS#1 parsing to extract the key type (one of the reasons for the **`public_key_info_decode()`**  function)

Reply:
OK, I guess it's unlikely that `"ML-DSA"` keys are of a (sub-)type we don't support (i.e. don't know the OID of - not sure if that could be an issue in the future, though). Would it make sense to check if the type is correct? That is, always decode the blob and then compare the type if it's not `KEY_ANY`? Or just always use the decoded type if it's known?